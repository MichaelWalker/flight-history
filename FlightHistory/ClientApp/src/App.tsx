import type { FunctionComponent} from "react";
import React, { useContext } from "react";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import { FlightsPage } from "./pages/Flights";
import { AirportsPage } from "./pages/Airports";
import { AircraftPage } from "./pages/Aircraft";
import { NotFoundPage } from "./pages/NotFound";
import { DashboardPage } from "./pages/Dashboard";
import { UserContext, UserContextProvider } from "./contexts/UserContext";
import { SignInPage } from "./pages/SignInPage";
import "./styles/reset.scss";
import { AirlinesPage } from "./pages/Airlines";
import { LoadingAnimation } from "./icons/loadingAnimation/LoadingAnimation";

export const App: FunctionComponent = () => {
    return (
        <UserContextProvider>
            <Router>
                <AppContent />
            </Router>
        </UserContextProvider>
    );
};

const AppContent: FunctionComponent = () => {
    const { loading, currentUser } = useContext(UserContext);

    if (loading) {
        return <LoadingAnimation size={"LARGE"} />;
    }

    if (!currentUser) {
        return <SignInPage />;
    }

    return (
        <Switch>
            <Route path={"/"} exact={true}>
                <DashboardPage />
            </Route>
            <Route path={"/flights"} exact={true}>
                <FlightsPage />
            </Route>
            <Route path={"/airports"} exact={true}>
                <AirportsPage />
            </Route>
            <Route path={"/aircraft"} exact={true}>
                <AircraftPage />
            </Route>
            <Route path={"/airlines"} exact={true}>
                <AirlinesPage />
            </Route>
            <Route path={"/"}>
                <NotFoundPage />
            </Route>
        </Switch>
    );
};
