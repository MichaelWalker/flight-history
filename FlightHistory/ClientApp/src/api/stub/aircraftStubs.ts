import type { StubApiResponse} from "./stubApiResponse";
import { stubItemList } from "./stubApiResponse";
import { toURL } from "../apiHelpers";
import type { Aircraft } from "../../models/aircraft";

export const aircraft: Aircraft[] = [
    { id: 1, registration: "VP-BWX", model: "767" },
    { id: 2, registration: "VP-BDN", model: "A319" },
];

export const AircraftStubs: StubApiResponse[] = [
    {
        url: "/api/aircraft",
        method: "GET",
        getResponseBody: (urlString: string): Response => {
            const url = toURL(urlString);
            const page = stubItemList(url, aircraft);
            return new Response(JSON.stringify(page));
        },
    },
];
