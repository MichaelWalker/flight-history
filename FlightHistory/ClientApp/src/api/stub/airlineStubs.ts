import type { StubApiResponse } from "./stubApiResponse";
import { stubItemList } from "./stubApiResponse";
import { toURL } from "../apiHelpers";
import { stubAirlineList } from "../../models/airline.testdata";

export const AirlineStubs: StubApiResponse[] = [
    {
        url: "/api/airlines",
        method: "GET",
        getResponseBody: (urlString: string): Response => {
            const url = toURL(urlString);
            const page = stubItemList(url, stubAirlineList);
            return new Response(JSON.stringify(page));
        },
    },
];
