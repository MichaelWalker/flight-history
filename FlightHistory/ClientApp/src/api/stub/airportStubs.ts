import type { StubApiResponse } from "./stubApiResponse";
import { stubItemList } from "./stubApiResponse";
import { toURL } from "../apiHelpers";
import { stubAirportList } from "../../models/airport.testdata";

export const AirportStubs: StubApiResponse[] = [
    {
        url: "/api/airports",
        method: "GET",
        getResponseBody: (urlString: string): Response => {
            const url = toURL(urlString);
            const page = stubItemList(url, stubAirportList, ["name", "code"]);
            return new Response(JSON.stringify(page));
        },
    },
];
