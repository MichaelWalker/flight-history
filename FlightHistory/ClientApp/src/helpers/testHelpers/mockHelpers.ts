import * as globals from "../globals";

export function mockUseSampleData(useSampleData: boolean) {
    jest.spyOn(globals, "useSampleData").mockReturnValue(useSampleData);
}

export function mockUseConsoleLogging(useConsoleLogging: boolean) {
    jest.spyOn(globals, "useConsoleLogging").mockReturnValue(useConsoleLogging);
}
