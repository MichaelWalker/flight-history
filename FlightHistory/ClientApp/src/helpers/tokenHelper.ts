import {Observable} from "./observable";
import {User} from "../models/user";

interface ParsedToken {
    expiry?: number;
    user: User;
}

export const accessToken = Observable.of<string | null>(null);

function parseToken(): ParsedToken | undefined {
    const token = accessToken.get();
    if (!token) {
        return;
    }

    const encodedPayload = token.split('.')[1];
    const decodedPayload = atob(encodedPayload);
    const data = JSON.parse(decodedPayload);

    return {
        expiry: tryParseExpiry(data),
        user: {
            name: data.name,
            email: data.email,
        }
    }
}

function tryParseExpiry(data: { exp?: number }): number | undefined {
    const expiry = data.exp;

    if (!expiry) {
        return;
    }

    return expiry * 1000;
}

export function accessTokenAppearsValid(): boolean {
    const payload = parseToken();

    if (!payload) {
        return false;
    }

    if (payload.expiry && payload.expiry < new Date().getTime()) {
        return false;
    }

    return true;
}

export function getCurrentUser(): User | undefined {
    return parseToken()?.user;
}