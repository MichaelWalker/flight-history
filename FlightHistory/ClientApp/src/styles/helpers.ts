interface RBG {
    red: number;
    green: number;
    blue: number;
}

export function rgba(hexColor: string, opacity: number): string {
    const { red, green, blue } = toRGB(hexColor);
    return `rgba(${red}, ${green}, ${blue}, ${opacity})`;
}

function toRGB(hexColor: string): RBG {
    const hex = hexColor.replace("#", "");

    return {
        red: parseHex(hex.substring(0, 2)),
        green: parseHex(hex.substring(2, 4)),
        blue: parseHex(hex.substring(4, 6)),
    };
}

function parseHex(hex: string): number {
    return parseInt(`0x${hex}`, 16);
}
