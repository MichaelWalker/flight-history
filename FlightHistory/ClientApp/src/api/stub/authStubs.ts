import type { StubApiResponse } from "./stubApiResponse";
import type { User } from "../../models/user";
import { generateTestAccessToken } from "../../helpers/stubTokenHelpers";
import { stubUser } from "../../models/user.testdata";
import { stubToken } from "../../models/token.testdata";

export let signedInUser: User | null = stubUser;

export function setSignedInUser(user: User | null) {
    signedInUser = user;
}

export const AuthStubs: StubApiResponse[] = [
    {
        url: "/api/auth/sign-in",
        method: "POST",
        getResponseBody: (url: string, data: string): Response => {
            const parsedData = JSON.parse(data) as {
                email: string;
                password: string;
            };
            const { email, password } = parsedData;

            if (email !== "user@sample.com" || password !== "password") {
                return {
                    ...Response.error(),
                    status: 401,
                };
            }

            signedInUser = stubUser;
            return new Response(
                JSON.stringify({
                    token: stubToken,
                }),
            );
        },
        allowUnauthorised: true,
    },
    {
        url: "/api/auth/refresh",
        method: "POST",
        getResponseBody: (): Response => {
            if (!signedInUser) {
                return {
                    ...Response.error(),
                    status: 401,
                };
            }
            return new Response(
                JSON.stringify({
                    token: generateTestAccessToken(signedInUser, 5),
                }),
            );
        },
        allowUnauthorised: true,
    },
    {
        url: "/api/auth/sign-out",
        method: "POST",
        getResponseBody: (): Response => {
            signedInUser = null;
            return new Response("");
        },
    },
];
