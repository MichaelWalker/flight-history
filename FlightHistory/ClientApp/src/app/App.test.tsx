import React from "react";
import { mockUseConsoleLogging, mockUseSampleData } from "../helpers/testHelpers/mockHelpers";
import { render, waitFor } from "@testing-library/react";
import { getAccessToken } from "../helpers/tokenHelper";
import { setSignedInUser, signedInUser } from "../api/stub/authStubs";
import { stubUser } from "../models/user.testdata";
import { App } from "./App";

describe("App", () => {
    const originalSignedInUser = signedInUser;

    beforeEach(() => {
        mockUseSampleData(true);
        mockUseConsoleLogging(false);
        getAccessToken().set(null);
    });

    afterEach(() => {
        jest.resetAllMocks();
        setSignedInUser(originalSignedInUser);
    });

    it("displays sign in page if user is not signed in", async () => {
        setSignedInUser(null);

        const { container, getByTestId } = render(<App />);

        expect(getByTestId("loading-animation")).toBeInTheDocument();
        expect(container).toMatchSnapshot("app-loading");

        await waitFor(() => expect(getByTestId("page-title").textContent).toBe("Sign In"));
        expect(container).toMatchSnapshot("sign-in-page");
    });

    it("displays dashboard if user is signed in", async () => {
        setSignedInUser(stubUser);

        const { container, getByTestId } = render(<App />);

        expect(getByTestId("loading-animation")).toBeInTheDocument();
        expect(container).toMatchSnapshot("app-loading");

        await waitFor(() => expect(getByTestId("page-title").textContent).toBe("Dashboard"));
        expect(container).toMatchSnapshot("dashboard-page");
    });
});
