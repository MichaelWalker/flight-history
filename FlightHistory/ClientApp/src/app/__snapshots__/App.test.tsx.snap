// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App displays dashboard if user is signed in: app-loading 1`] = `
<div>
  <svg
    class="sc-gKAaRy iMvhJB"
    data-testid="loading-animation"
    preserveAspectRatio="xMidYMid"
    viewBox="0 0 100 100"
  >
    <circle
      class="sc-iCoGMd iEiTVC"
      cx="50"
      cy="50"
      r="32"
      stroke-dasharray="50.26548245743669 50.26548245743669"
    >
      <animatetransform
        attributeName="transform"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        type="rotate"
        values="0 50 50;360 50 50"
      />
    </circle>
    <circle
      class="sc-iCoGMd nHcLQ"
      cx="50"
      cy="50"
      r="26"
      stroke-dasharray="40.840704496667314 40.840704496667314"
      stroke-dashoffset="40.840704496667314"
    >
      <animatetransform
        attributeName="transform"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        type="rotate"
        values="0 50 50;-360 50 50"
      />
    </circle>
  </svg>
</div>
`;

exports[`App displays dashboard if user is signed in: dashboard-page 1`] = `
<div>
  <div
    class="sc-bdnxRM jopmEo"
  >
    <div
      class="sc-bdnxRM eXUoTH"
    >
      <a
        class="sc-fujyAs JsHRD"
        css="
    ,
        font-size: ,1.25rem,;
        line-height: ,2rem,;
        font-weight: ,300,;
        font-family: \\"Ubuntu\\", sans-serif;
        color: ,var(--color-foreground),;
        text-transform: ,uppercase,;
    ,
    margin: ,1rem, ,2rem, ,6rem,;
    padding-bottom: ,1rem,;
    text-align: center;
"
        href="/"
      >
        Flight History
      </a>
      <nav
        class="nav"
      >
        <a
          aria-current="page"
          class="navItem active"
          href="/"
        >
          <div
            class="iconContainer"
          >
            <svg
              class="sc-gKAaRy sJbFD"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M4 13h6c.55 0 1-.45 1-1V4c0-.55-.45-1-1-1H4c-.55 0-1 .45-1 1v8c0 .55.45 1 1 1zm0 8h6c.55 0 1-.45 1-1v-4c0-.55-.45-1-1-1H4c-.55 0-1 .45-1 1v4c0 .55.45 1 1 1zm10 0h6c.55 0 1-.45 1-1v-8c0-.55-.45-1-1-1h-6c-.55 0-1 .45-1 1v8c0 .55.45 1 1 1zM13 4v4c0 .55.45 1 1 1h6c.55 0 1-.45 1-1V4c0-.55-.45-1-1-1h-6c-.55 0-1 .45-1 1z"
              />
            </svg>
          </div>
          Dashboard
        </a>
        <a
          class="navItem"
          href="/flights"
        >
          <div
            class="iconContainer"
          >
            <svg
              class="sc-gKAaRy sJbFD"
              viewBox="0 0 24 24"
            >
              <g>
                <path
                  d="M0,0h24v24H0V0z"
                  fill="none"
                />
              </g>
              <g>
                <path
                  d="M20.19,4H4C2.9,4,2.01,4.9,2.01,6v4C3.11,10,4,10.9,4,12s-0.89,2-2,2v4c0,1.1,0.9,2,2,2h16c1.1,0,2-0.9,2-2V6 C22,4.9,21.19,4,20.19,4z M17.73,13.3l-8.49,2.26c-0.22,0.06-0.45-0.04-0.56-0.23l-1.12-1.95c-0.18-0.3-0.01-0.69,0.32-0.78h0 c0.16-0.04,0.34-0.01,0.47,0.1l1.05,0.82l2.39-0.64L9.9,9.6c-0.26-0.44-0.02-1.01,0.47-1.15l0,0c0.26-0.07,0.54,0,0.74,0.18 l3.69,3.44l2.44-0.65c0.51-0.14,1.04,0.17,1.18,0.68C18.55,12.62,18.25,13.15,17.73,13.3z"
                />
              </g>
            </svg>
          </div>
          Flights
        </a>
        <a
          class="navItem"
          href="/airports"
        >
          <div
            class="iconContainer"
          >
            <svg
              class="sc-gKAaRy sJbFD"
              viewBox="0 0 24 24"
            >
              <g>
                <rect
                  fill="none"
                  height="24"
                  width="24"
                  y="0"
                />
              </g>
              <g>
                <path
                  d="M12,2c-4.2,0-8,3.22-8,8.2c0,3.18,2.45,6.92,7.34,11.23c0.38,0.33,0.95,0.33,1.33,0 C17.55,17.12,20,13.38,20,10.2C20,5.22,16.2,2,12,2z M12,12c-1.1,0-2-0.9-2-2c0-1.1,0.9-2,2-2c1.1,0,2,0.9,2,2 C14,11.1,13.1,12,12,12z"
                />
              </g>
            </svg>
          </div>
          Airports
        </a>
        <a
          class="navItem"
          href="/aircraft"
        >
          <div
            class="iconContainer"
          >
            <svg
              class="sc-gKAaRy sJbFD"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M21 14.58c0-.36-.19-.69-.49-.89L13 9V3.5c0-.83-.67-1.5-1.5-1.5S10 2.67 10 3.5V9l-7.51 4.69c-.3.19-.49.53-.49.89 0 .7.68 1.21 1.36 1L10 13.5V19l-1.8 1.35c-.13.09-.2.24-.2.4v.59c0 .33.32.57.64.48L11.5 21l2.86.82c.32.09.64-.15.64-.48v-.59c0-.16-.07-.31-.2-.4L13 19v-5.5l6.64 2.08c.68.21 1.36-.3 1.36-1z"
              />
            </svg>
          </div>
          Aircraft
        </a>
        <a
          class="navItem"
          href="/airlines"
        >
          <div
            class="iconContainer"
          >
            <svg
              class="sc-gKAaRy sJbFD"
              viewBox="0 0 24 24"
            >
              <rect
                fill="none"
                height="24"
                width="24"
                y="0"
              />
              <path
                d="M19.59,4h-5.01c-0.99,0-1.91,0.49-2.47,1.3L2,20h17l2.56-13.63C21.79,5.14,20.84,4,19.59,4z M14.5,14 c-1.38,0-2.5-1.12-2.5-2.5c0-1.38,1.12-2.5,2.5-2.5s2.5,1.12,2.5,2.5C17,12.88,15.88,14,14.5,14z"
              />
            </svg>
          </div>
          Airlines
        </a>
      </nav>
    </div>
    <div
      class="sc-bdnxRM UA-Dih"
    >
      <header
        class="sc-gtsrHT eQLWfg"
      >
        <h1
          class="sc-dlnjwi jTPWQM"
          data-testid="page-title"
        >
          Dashboard
        </h1>
      </header>
      <main
        class="sc-eCApnc krkCEB"
      >
        Hello!
      </main>
    </div>
  </div>
</div>
`;

exports[`App displays sign in page if user is not signed in: app-loading 1`] = `
<div>
  <svg
    class="sc-gKAaRy iMvhJB"
    data-testid="loading-animation"
    preserveAspectRatio="xMidYMid"
    viewBox="0 0 100 100"
  >
    <circle
      class="sc-iCoGMd iEiTVC"
      cx="50"
      cy="50"
      r="32"
      stroke-dasharray="50.26548245743669 50.26548245743669"
    >
      <animatetransform
        attributeName="transform"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        type="rotate"
        values="0 50 50;360 50 50"
      />
    </circle>
    <circle
      class="sc-iCoGMd nHcLQ"
      cx="50"
      cy="50"
      r="26"
      stroke-dasharray="40.840704496667314 40.840704496667314"
      stroke-dashoffset="40.840704496667314"
    >
      <animatetransform
        attributeName="transform"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        type="rotate"
        values="0 50 50;-360 50 50"
      />
    </circle>
  </svg>
</div>
`;

exports[`App displays sign in page if user is not signed in: sign-in-page 1`] = `
<div>
  <main
    class="main"
  >
    <section
      class="section"
    >
      <h1
        class="title"
        data-testid="page-title"
      >
        Sign In
      </h1>
      <form>
        <label
          class="field"
        >
          <span
            class="label"
          >
            Email
          </span>
          <input
            class="input"
            required=""
            type="email"
            value=""
          />
        </label>
        <label
          class="field"
        >
          <span
            class="label"
          >
            Password
          </span>
          <input
            class="input"
            required=""
            type="password"
            value=""
          />
        </label>
        <button
          class="button"
          disabled=""
          type="submit"
        >
          Sign In
        </button>
        <div
          class="errorMessage"
        />
      </form>
    </section>
  </main>
</div>
`;
